accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,484 @@
+    /// ObservableGauge is an observable Instrument that reports non-additive value(s)
+    /// when the instrument is being observed.
+    /// e.g. the current room temperature
+    /// </summary>
+    public sealed class ObservableGauge<T> : ObservableInstrument<T> where T : unmanaged
+    {
+        protected override IEnumerable<Measurement<T>> Observe() { throw null; }
+    }
+}
+```
+
+### MeterListener
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A delegate to represent the callbacks signatures used in the listener.
+    /// </summary>
+    public delegate void MeasurementCallback<T>(
+                            Instrument instrument,
+                            T measurement,
+                            ReadOnlySpan<KeyValuePair<string, object?>> tags,
+                            object? state);
+
+
+    /// <summary>
+    /// The listener class can be used to listen to observable and non-observable instrument
+    /// recorded measurements.
+    /// </summary>
+    public class MeterListener : IDisposable
+    {
+        /// <summary>
+        /// Simple constructor
+        /// </summary>
+        public MeterListener() { throw null;  }
+
+        /// <summary>
+        /// Callbacks to get notification when an instrument is published
+        /// </summary>
+        public Action<Instrument, MeterListener>? InstrumentPublished { get; set; }
+
+        /// <summary>
+        /// Callbacks to get notification when stopping the measurement on some instrument
+        /// this can happen when the Meter or the Listener is disposed of. Or calling Stop()
+        /// on the listener.
+        /// </summary>
+        // This need some clarification
+        public Action<Instrument, object?>? MeasurementsCompleted { get; set; }
+
+        /// <summary>
+        /// Start listening to a specific instrument measurement recording.
+        /// </summary>
+        public void EnableMeasurementEvents(Instrument instrument, object? state = null) { throw null; }
+
+        /// <summary>
+        /// Stop listening to a specific instrument measurement recording.
+        /// </summary>
+        public void DisableMeasurementEvents(Instrument instrument) { throw null; }
+
+        /// <summary>
+        /// Set a callback for a specific numeric type to get the measurement recording notification
+        /// from all instruments which enabled listened to and was created with the same specified
+        /// numeric type. If a measurement of type T is recorded and a callback of type T is registered, that callback is used. If there is no callback for type T but there is a callback for type object, the measured value is boxed and reported via the object typed callback. If there is neither type T callback nor object callback then the measurement will not be reported.
+        /// </summary>
+        public void SetMeasurementEventCallback<T>(MeasurementCallback<T>? measurementCallback) { throw null; }
```

Comment:
Given MeasurementCallback is a delegate, do you intend to only allow one callback, or multiple?

Reply:
It will be one callback per numeric type.